src/asl.ts

```diff
@@ -2129,33 +2319,152 @@ export class ASL {
           ErrorCodes.Cannot_perform_arithmetic_or_bitwise_computations_on_variables_in_Step_Function,
           `Step Function does not support operator ${expr.op}`
         );
-      } else if (
-        expr.op === "instanceof" ||
-        // https://github.com/functionless/functionless/issues/393
-        expr.op === "??=" ||
-        expr.op === "||="
-      ) {
+      } else if (expr.op === "instanceof") {
         throw new SynthError(
           ErrorCodes.Unsupported_Feature,
           `Step Function does not support ${expr.op} operator`
         );
-      }
-      assertNever(expr.op);
-    } else if (isAwaitExpr(expr)) {
-      return this.eval(expr.expr);
-    } else if (isTypeOfExpr(expr)) {
-      return this.evalExpr(expr.expr, (exprOutput) => {
-        if (ASLGraph.isLiteralValue(exprOutput)) {
-          return {
-            value: typeof exprOutput.value,
-            containsJsonPath: false,
-          };
-        }
-
-        const tempHeap = this.newHeapVariable();
+      } else if (expr.op === "&&" || expr.op === "||" || expr.op === "??") {
+        return this.evalExpr(expr.left, (leftOutput) => {
+          const right = this.eval(expr.right);
```

Comment:
So, if i undersetand correctly, eval(left) may reeturns a JsonPath, Condition or LiteralValue. If it returns a LiteralValue, does that mean no extra state was added to the machine?

Reply:
Not necessarily, `(await Task(), true)` would add some states and return a `LiteralValue`. We shouldn't make assumption about the states based on the `ASLGraph.Output` type. Always assume that `eval` can return states or just outputs.

However, if `ASLGraph.isAslGraphOutput(this.eval(expr))` is `true`, then you know no states were added. I use this fact here: https://github.com/functionless/functionless/blob/sfn_improve_binary_op/src/asl.ts#L2404.

`right` here isn't evaluated with `evalExpr` it uses `eval` which returns the entire node and output, so I know if `right` is a LiteralValue then there is no other work to do (and no short circuiting risks).